RecoLocalCalo/EcalRecProducers/plugins/alpaka/AmplitudeComputationCommonKernels.h

```diff
@@ -0,0 +1,491 @@
+#include "HeterogeneousCore/AlpakaInterface/interface/traits.h"
+#include "RecoLocalCalo/EcalRecProducers/interface/EigenMatrixTypes_gpu.h"
+
+#include "DeclsForKernels.h"
+#include "KernelHelpers.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE::ecal::multifit {
+
+  ///
+  /// assume kernel launch configuration is
+  /// (MAXSAMPLES * nchannels, blocks)
+  /// TODO: is there a point to split this kernel further to separate reductions
+  ///
+  class Kernel_prep_1d_and_initialize {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                  EcalDigiDeviceCollection::ConstView digisDevEB,
+                                  EcalDigiDeviceCollection::ConstView digisDevEE,
+                                  EcalUncalibratedRecHitDeviceCollection::View uncalibRecHitsEB,
+                                  EcalUncalibratedRecHitDeviceCollection::View uncalibRecHitsEE,
+                                  EcalMultifitConditionsDevice::ConstView conditionsDev,
+                                  ::ecal::multifit::SampleVector* amplitudes,
+                                  ::ecal::multifit::SampleGainVector* gainsNoise,
+                                  bool* hasSwitchToGain6,
+                                  bool* hasSwitchToGain1,
+                                  bool* isSaturated,
+                                  char* acState,
+                                  ::ecal::multifit::BXVectorType* bxs,
+                                  bool const gainSwitchUseMaxSampleEB,
+                                  bool const gainSwitchUseMaxSampleEE) const {
+      constexpr bool dynamicPedestal = false;  //---- default to false, ok
+      constexpr auto nsamples = EcalDataFrame::MAXSAMPLES;
+      constexpr int sample_max = 5;
+      constexpr int full_pulse_max = 9;
+      auto const offsetForHashes = conditionsDev.offsetEE();
+
+      auto const nchannelsEB = digisDevEB.size();
+      auto const nchannelsEE = digisDevEE.size();
+      auto const nchannels = nchannelsEB + nchannelsEE;
+      auto const totalElements = nchannels * nsamples;
+
+      auto const elemsPerBlock = alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[0u];
+
+      char* shared_mem = alpaka::getDynSharedMem<char>(acc);
+      auto* shr_hasSwitchToGain6 = reinterpret_cast<bool*>(shared_mem);
+      auto* shr_hasSwitchToGain1 = shr_hasSwitchToGain6 + elemsPerBlock;
+      auto* shr_hasSwitchToGain0 = shr_hasSwitchToGain1 + elemsPerBlock;
+      auto* shr_isSaturated = shr_hasSwitchToGain0 + elemsPerBlock;
+      auto* shr_hasSwitchToGain0_tmp = shr_isSaturated + elemsPerBlock;
+      auto* shr_counts = reinterpret_cast<char*>(shr_hasSwitchToGain0_tmp) + elemsPerBlock;
+
+      // pulse shape template
+      auto const* shapes_in = reinterpret_cast<const EcalPulseShape*>(conditionsDev.pulseShapes()->data());
```

Comment:
The "Notes" section in https://en.cppreference.com/w/cpp/language/reinterpret_cast has

> Performing a class member access that designates a non-static data member or a non-static member function on a glvalue that does not actually designate an object of the appropriate type - such as one obtained through a `reinterpret_cast` - results in undefined behavior

and shows an example of very similar case as `PulseShapeArray` vs. `EcalPulseShape`, and there accessing the member data via the casted pointer is said to be undefined behavior. This is despite of (based on what I played in compiler explorer)
* `std::array<float, EcalPulseShape::TEMPLATESAMPLES>` is a standard layout class
  * i.e. a pointer to the first data member (returned by `data()`) can be converted to a pointer to the `array` object, and vice versa
* `EcalPulseShape` is also a standard layout class
* `std::array<float, EcalPulseShape::TEMPLATESAMPLES>` and `EcalPulseShape` have compatible layouts
   * e.g. C++20 `std::is_layout_compatible_v<std::array<...>, EcalPulseShape> == true`

----



I saw `shapes_in` being used only at line 294
```cpp
auto shape_value = shapes_in[hashedId].pdfval[full_pulse_max - 7];
```
Maybe it could be replaced with e.g.
```cpp
auto shape_value = conditionsDev.pulseShapes()[hashedId][full_pulse_max - 7];
```
 to avoid the `reinterpret_cast`?

Reply:
This seems to work, yes. It compiles and runs at least.